core/src/main/java/com/linecorp/armeria/client/limit/ConcurrencyLimit.java

```diff
@@ -54,6 +56,50 @@ static ConcurrencyLimitBuilder builder(int maxConcurrency) {
         return new ConcurrencyLimitBuilder(maxConcurrency == MAX_VALUE ? 0 : maxConcurrency);
     }
 
+    /**
+     * Returns a newly-created {@link ConcurrencyLimit} with the specified {@code maxConcurrency}.
+     *
+     * <pre>
+     *     ConcurrencyLimit.of(new DynamicLimit());
+     *
+     *     class DynamicLimit extends SettableLimit {
+     *         public DynamicLimit() {
+     *             super(defaultValue);
+     *             dogma.watcher().watch((rev, value) -> {
+     *                 set(Integer.parse(value));
+     *             });
+     *         }
+     *     }
+     * </pre>
+     *
+     * @param maxConcurrency the settable maximum number of concurrent active requests.
+     */
+    static ConcurrencyLimit of(SettableLimit maxConcurrency) {
+        return builder(maxConcurrency).build();
+    }
+
+    /**
+     * Returns a new {@link ConcurrencyLimitBuilder} with the specified {@code maxConcurrency}.
+     *
+     * <pre>
+     *     ConcurrencyLimit.builder(new DynamicLimit());
+     *
+     *     class DynamicLimit extends SettableLimit {
+     *         public DynamicLimit() {
+     *             super(defaultValue);
+     *             dogma.watcher().watch((rev, value) -> {
+     *                 set(Integer.parse(value));
+     *             });
+     *         }
+     *     }
+     * </pre>
+     *
+     * @param maxConcurrency the settable maximum number of concurrent active requests.
+     */
+    static ConcurrencyLimitBuilder builder(SettableLimit maxConcurrency) {
+        return new ConcurrencyLimitBuilder(maxConcurrency);
```

Comment:
If we use `SettableLimit` for `maxConcurrency`, we force the `ConcurrencyLimit` to use the `AtomicInteger` variable which takes a bit of cost.
How about just using `Supplier<Integer>` and providing `SettableLimit` as a utility class? so that
- users can just use an integer parameter as the `ConcurrencyLimit` with no overhead as it is.
- we can give more freedom to users to implement the `Supplier`.
- we can also provide a convenient way to cache the value via `SettableLimit` when it has heavy calculating logic for the limit.

Reply:
One more thing to consider is how about using `Function<ClientRequestContext, Integer>` instead of `Supplier<Integer>`?
Users might want to calculate the `concurrencyLimit` based on the information from the `ClientRequestContext`. 🤔